src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,233 @@
+        if gets.to_s.lowercase == "y"
+          CLI.generate_config
         end
+        exit 1
       end
+    rescue ex : KeyError
+      error "Error in watch configuration. #{ex.message}"
+      exit 1
     end
 
-    private def scan_files
+    private def run_watcher(watch_config)
+      begin
+        server_config = watch_config["server"]
+        spawn watcher("server", server_config["files"], server_config["commands"])
+      rescue ex
+        error "Error in watch configuration. #{ex.message}"
+        exit 1
+      end
+      watch_config.each do |key, value|
+        next if key == "client" || key == "server"
+        files = value["files"]
+        commands = value["commands"]
+        @notify_counter += 1
+        spawn watcher(key, files, commands)
+      end
+      @notify_counter_channel.send @notify_counter
+      @notify_counter = @notify_counter_channel.receive
+      sleep
+    end
+
+    private def watcher(key, files, commands)
+      if key != "server"
+        @notify_channel.receive
+      end
+      if files.empty?
+        commands.each do |command|
+          run_command(command, key)
+        end
+      else
+        loop do
+          scan_files(key, files, commands)
+          @watch_running = true if key == "server"
+          sleep 1
+        end
+      end
+    end
+
+    private def scan_files(key, files, commands)
       file_counter = 0
-      Dir.glob(files) do |file|
-        timestamp = get_timestamp(file)
-        if FILE_TIMESTAMPS[file]? != timestamp
-          if @app_running
-            log "File changed: #{file.colorize(:light_gray)}"
-          end
-          FILE_TIMESTAMPS[file] = timestamp
-          file_counter += 1
+      @file_watcher.scan_files(files) do |file|
+        if @watch_running
+          debug "File changed: #{file}"
         end
+        file_counter += 1
       end
       if file_counter > 0
-        log "Watching #{file_counter} files (server reload)..."
-        start_app
+        debug "Watching #{file_counter} #{key} files"
+        kill_processes(key)
+        commands.each do |command|
+          if key == "server" && command == commands.first?
+            run_build_command(command, commands)
+          else
+            run_command(command, key)
+          end
+        end
       end
     end
 
-    private def stop_all_processes
-      log "Terminating app #{project_name}..."
-      @processes.each do |process|
-        process.kill unless process.terminated?
+    private def check_directories
+      Dir.mkdir_p("bin")
+      if !Dir.exists?("lib")
+        error "You need to install dependencies first, execute `shards install`"
+        exit 1
       end
-      processes.clear
     end
 
-    private def create_all_processes
-      process = create_watch_process
-      @processes << process if process.is_a? Process
-      unless @npm_process
-        create_npm_process
-        @npm_process = true
+    private def run_build_command(command, commands)
```

Comment:
Should be its own class/stuct

Reply:
This won't be required, I just gonna use `Helpers.run` to run processes :+1: